### PR TITLE
With workbox-window

### PR DESCRIPTION
**Pull request by azizhk**
The workbox webpack plugin has significantly improved from #2549 and does not require us to be as verbose. This is using the alpha release of workbox v5, we can merge after a stable release.

This example covers:
- [x] Service Worker with Typescript
- [x] Service Worker Bundling (Provided by the workbox webpack plugin)
- [x] Precaching of client side assets when service worker starts
- [x] Communication to user that new update is available.
- [x] Reload all existing tabs when Service Worker is updated to avoid state mismatch.
- [x] Update service worker when user navigates (by clicking on link or some action)
- [x] Update service worker when user refreshes the page
- [x] Use `assetPrefix` when assets are uploaded to CDN
- [x] Precache HTML assets as well when in export mode.
- [x] Add a few more sensible guidelines from https://redfin.engineering/how-to-fix-the-refresh-button-when-using-service-workers-a8e27af6df68

I know that an example with workbox can be more concise but I wanted the example to contain a good set of defaults so that they don't come back to bite developers as it's very easy to mis-configure Service Workers.

I've not taken "adding an entry to main.js" route because I wanted to communicate with the react app when new versions available.
CC: @ragingwind would like your review.
There is also https://github.com/hanford/next-offline which also has examples of using the v5 alpha release.

### CODE REVIEW

File: examples/with-workbox/components/ServiceWorkerRegistrar.tsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          { message: 'Service Worker Installed' },
+          showSnackMessage
+        )
+      )
+    })
+    wb.addEventListener('waiting', this.onSWWaiting)
+
+    Router.events.on('routeChangeStart', e => {
+      if (this.state.swUpdateWaiting && this.wb) {
+        window.location.reload()
```
Comment (azizhk) [inline examples/with-workbox/components/ServiceWorkerRegistrar.tsx]: Here will have to skipWaiting and then fetch the new route.

File: examples/with-workbox/components/ServiceWorkerRegistrar.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+          showSnackMessage
+        )
+      )
+    })
+    wb.addEventListener('waiting', this.onSWWaiting)
+    setTimeout(this.checkForUpdates, HOUR_IN_MS)
+  }
+
+  private onSWWaiting = () => {
+    Router.events.on('routeChangeStart', this.onRouteChange)
```
Comment (azizhk) [inline examples/with-workbox/components/ServiceWorkerRegistrar.tsx]: Can't cancel navigation: https://github.com/zeit/next.js/issues/2476

File: examples/with-workbox/README.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+yarn dev
+```
+
+Deploy it to the cloud with [now](https://zeit.co/now) ([download](https://zeit.co/download))
+
+```bash
+now
+```
+
+## To remove Typescript
```
Comment (lfades) [inline examples/with-workbox/README.md]: Let's not include this

File: examples/with-workbox/README.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+```
+"build": "next build && next export && workbox injectManifest workbox-config.js"
+"start": "serve ./out"
+```
+
+In sw.ts modify `precacheAndRoute(self.__WB_MANIFEST);` to `precacheAndRoute(self.__WB_MANIFEST.concat(self.__WB_INJECT_MANIFEST));` (You will have to modify type definition).
+
+## The idea behind the example
+
+Implementation of https://redfin.engineering/how-to-fix-the-refresh-button-when-using-service-workers-a8e27af6df68
```
Comment (lfades) [inline examples/with-workbox/README.md]: Remove this part, the link is already shared at the end

File: examples/with-workbox/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+- Reload all existing tabs when Service Worker is updated to avoid state mismatch.
+- Update service worker when user navigates (by clicking on link or some action)
+- Update service worker when user refreshes the page
+- How to use assetPrefix when assets are uploaded to CDN
+
+Further Reading:
+
+- https://redfin.engineering/how-to-fix-the-refresh-button-when-using-service-workers-a8e27af6df68
+- next-offline: https://github.com/hanford/next-offline
+- Udacity Course: https://classroom.udacity.com/courses/ud899
```
Comment (lfades) [inline examples/with-workbox/README.md]: This takes me to a login page, share a public URL that won't require a sign up / sign in

**Comment by Timer** [comment]
Hey @azizhk!

We really appreciate the work that's gone into this pull request, but I don't think we're going to be accepting a Workbox example right now.

We’re currently collaborating with Google Chrome and will potentially have first-class support for Workbox in the future. It’ll be better if we postpone this so we don’t need to worry about backwards compatibility with this example.

We appreciate your understanding. Users can still find this by searching the issue tracker!

**Comment by azizhk** [comment]
Cool, no issues.
Better to have the original workbox developers handle things. Service workers are easily misconfigured. Create React App had a lot of issues. I totally understand postponing to make it perfect.

**Comment by cansin** [comment]
Any update on that first-class support @Timer ?

**Comment by RajNandigalla** [comment]
@azizhk  nice example , i finally got it working , i was missing manifestTransform filtering those helped

**Pull request closed without merging**